### PR TITLE
Start workspace from devfile with private repo

### PR DESCRIPTION
**Pull request by svor**
### What does this PR do?
If a repository described in a devfile is private, the project will be cloned by using a terminal widget where Username and Password will be asked.
private_repo

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16401

### CODE REVIEW

File: plugins/workspace-plugin/src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    } catch (e) {
+        throw new Error('Unable to get list workspace containers. Cause: ' + e);
+    }
+    return containers;
+}
+
+export async function execInTerminal(directory: string, locationURI: string, ...args: string[]): Promise<void> {
+    const editorContainerName = await getEditorContainerName() as string;
+    try {
+        const shellArgs: string[] = ['-c', 'git'].concat(args);
```
Comment (RomanNikitenko) [inline plugins/workspace-plugin/src/utils.ts:60]: Should `git` come as arg?
I mean it's `utils` and `execInTerminal` function, so why `git` is hard coded here?

File: plugins/workspace-plugin/src/git.ts [hunk 2]
```diff
@@ -41,17 +41,15 @@ export async function getRemoteURL(remote: string, projectPath: string): Promise
...
     await initRepository(projectPath);
     // Enable sparse checkout feature
     await setConfig(projectPath, 'core.sparsecheckout', 'true');
     // Write sparse checkout directory
     const gitInfoFolderPath = path.join(projectPath, '.git/info/');
     fs.ensureDirSync(gitInfoFolderPath);
     fs.writeFileSync(path.join(gitInfoFolderPath, 'sparse-checkout'), sparseCheckoutDirectory);
-    // Add remote, pull changes and create the selected directory content
-    await execGit(projectPath, 'remote', 'add', '-f', 'origin', repositoryUri);
-    await execGit(projectPath, 'pull', 'origin', commitReference);
```
Comment (vitaliy-guliy) [inline plugins/workspace-plugin/src/git.ts:54]: Why did you remove sparse checkout feature?

File: plugins/workspace-plugin/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import { che as cheApi } from '@eclipse-che/api';
+import * as che from '@eclipse-che/plugin';
+import * as theia from '@theia/plugin';
+import { execute } from './exec';
+
+export async function getEditorContainerName(): Promise<string | undefined> {
```
Comment (vitaliy-guliy) [inline plugins/workspace-plugin/src/utils.ts:16]: Could you add some comments?

File: plugins/workspace-plugin/src/utils.ts (hunk 1 above)
Comment (vitaliy-guliy) [reply plugins/workspace-plugin/src/utils.ts:60]: Indeed. I think it should be possible to run `git` instead of `sh` in the following code.

File: plugins/workspace-plugin/src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    return containers;
+}
+
+export async function execInTerminal(directory: string, locationURI: string, ...args: string[]): Promise<void> {
+    const editorContainerName = await getEditorContainerName() as string;
+    try {
+        const shellArgs: string[] = ['-c', 'git'].concat(args);
+        const terminalOptions: theia.TerminalOptions = {
+            cwd: directory,
+            shellPath: 'sh',
```
Comment (vitaliy-guliy) [inline plugins/workspace-plugin/src/utils.ts:63]: Could you try to run `git`?

**Comment by benoitf** [comment]
`git clone` command provided by the `git plug-in` is asking login/password in the UI by default, so there is no way to reuse that flow?

File: plugins/workspace-plugin/src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        const terminal = theia.window.createTerminal(terminalOptions);
+        terminal.show();
+        await terminal.processId;
+    } catch (error) {
+        console.error(`Couldn't clone ${locationURI}: ${error.message}`);
+        throw new Error(`Couldn't clone ${locationURI}: ${error.message}`);
+    }
+}
+
+export async function checkRepoAccess(locationURI: string): Promise<boolean> {
```
Comment (vitaliy-guliy) [inline plugins/workspace-plugin/src/utils.ts:81]: If you test for 401, then the name of the method should be changed something like `isRepositoryPrivate`.
Btw, network errors are not handled here. The method will always return false for public repository and when error.

File: plugins/workspace-plugin/src/utils.ts (hunk 1 above)
Comment (RomanNikitenko) [reply plugins/workspace-plugin/src/utils.ts:60]: which comes as arg and is not hard coded
 :wink:

btw, the function executes a command in editor's container, so it should be defined in function name or the function should have the corresponding docs/comment or a container name should come as arg as well

**Comment by vitaliy-guliy** [comment]
@svor could you add the devfile to the PR description. With this title it looks like that devfile is placed somewhere on the private repo and you try to create a workspace from this devfile.

**Comment by svor** [comment]
@benoitf 

> git clone command provided by the git plug-in is asking login/password in the UI by default, so there is no way to reuse that flow?

We can call `git.clone` command with git URL param and it will clone the project with asking login/password in the UI. But if the devfile contains information about branch, folder's name where project should be located or if there is `sparseCheckoutDir` attribute, in that case vscode-git won't help.

**Comment by vinokurig** [comment]
I would suggest to not to use terminal to get the credentials. Instead you can check if the repository is private: https://stackoverflow.com/a/54964584 and if so show an input popup to fetch the credentials and insert them to the clone link.

**Comment by svor** [comment]
@benoitf @vinokurig @vitaliy-guliy @RomanNikitenko thx for your review, I'm working on providing another solution, will update the PR soon

**Comment by svor** [comment]
After last changes user will see the view:
clone_private2

FYI all askpass*.* files where ported form https://github.com/che-incubator/vscode-git 

@benoitf @vinokurig @vitaliy-guliy @RomanNikitenko could you take another look at this PR

File: plugins/workspace-plugin/src/askpass-main.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
+// tslint:disable:file-header
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+// Based on: https://github.com/che-incubator/vscode-git/blob/master/src/askpass-main.ts.
```
Comment (vinokurig) [inline plugins/workspace-plugin/src/askpass-main.ts:6]: May be we should use a link to original microsoft repository?

Comment (svor) [reply plugins/workspace-plugin/src/askpass-main.ts]: I think it does make sense. I've changed references. Do you know if we need to create some CQ for that?

Comment (vinokurig) [reply plugins/workspace-plugin/src/askpass-main.ts]: **AFAIK** we do need a CQ for code pieces.

**Comment by svor** [comment]
@benoitf @ericwill do you have any objections to merging this PR?

**Comment by svor** [comment]
I've created a CQ for ported code https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22087

**Comment by svor** [comment]
Looks like the CQ was resolved https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22087

**Comment by ericwill** [comment]
No objections from me, please merge. :+1:

**Comment by mukulk2020** [comment]
Could you please share the documentation link so that i can use my private repo inside devfile . @svor Please help me on this ..

**Comment by svor** [comment]
@mukulk2020 try to configure OAuth 2 for GitHub: https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-github/

**Comment by mukulk2020** [comment]
@svor I tried the same and configured in my ooenshift cluster ..but it is still asking for username and password of github ...

**Comment by svor** [comment]
@mukulk2020 does the devfile locate in your private repository with the project or it has another location?

**Comment by mukulk2020** [comment]
> @mukulk2020 does the devfile locate in your private repository with the project or it has another location?

With in the project only under root .

**Comment by mukulk2020** [comment]
Actually it is not picking the devfile as well from dev space dashboard while creating workspace

**Comment by svor** [comment]
which version of Che/DevSpaces are you using?

**Comment by mukulk2020** [comment]
> which version of Che/DevSpaces are you using?

devspacesoperator.v3.1.0-0.1661879180.p
schemaVersion: 2.1.0

**Comment by svor** [comment]
@vinokurig as I know if the OAuth was configured, che should not ask for username and password. Any advice here?

**Comment by vinokurig** [comment]
Che shouldn't ask git credentials if the private repository mentioned in the devfile has an access for the user which authenticated the factory.

**Comment by mukulk2020** [comment]
@vinokurig @svor : FYI
OCP Version: 4.10.22
Installed Dev Spaces through operatorHub.

image

CheCluster

image

Created the secret 
image

But it is still asking for username/password when workspace is started .

image

**Comment by vinokurig** [comment]
@mukulk2020 
Cold you please enable the `log DEBUG` mode and share the logs from the `devspaces` pod by applying:
```YAML
spec:
  components:
    cheServer:
      debug: true
      logLevel: debug
```
In the `Installed Operators` -> `Red Hat OpenShift Dev Spaces` -> `Red Hat OpenShift Dev Spaces instance Specification` -> `devspaces` -> `YAML` tab.

Please check the logs file before sharing, it may contain some sensitive data like tokens and OAuth secrets.

**Comment by mukulk2020** [comment]
2022-09-30 07:45:09,014[nio-8080-exec-2]  [DEBUG] [o.a.t.w.server.WsFrameServer 173]    - WebSocket frame received. fin [true], rsv [4], OpCode [1], payload length [59]
2022-09-30 07:45:09,025[nio-8080-exec-1]  [DEBUG] [a.c.w.i.BasicWebSocketEndpoint 144]  - Error: 
java.io.EOFException: null
	at org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper.fillReadBuffer(NioEndpoint.java:1301)
	at org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper.read(NioEndpoint.java:1188)
	at org.apache.tomcat.websocket.server.WsFrameServer.onDataAvailable(WsFrameServer.java:75)
	at org.apache.tomcat.websocket.server.WsFrameServer.doOnDataAvailable(WsFrameServer.java:183)
	at org.apache.tomcat.websocket.server.WsFrameServer.notifyDataAvailable(WsFrameServer.java:162)
	at org.apache.tomcat.websocket.server.WsHttpUpgradeHandler.upgradeDispatch(WsHttpUpgradeHandler.java:157)
	at org.apache.coyote.http11.upgrade.UpgradeProcessorInternal.dispatch(UpgradeProcessorInternal.java:60)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:59)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:872)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1705)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
2022-09-30 07:45:09,033[nio-8080-exec-1]  [DEBUG] [o.a.tomcat.websocket.WsSession 175]  - Failed to send close message for session [1] to remote endpoint
java.io.IOException: java.io.IOException: Broken pipe
	at org.apache.tomcat.websocket.WsRemoteEndpointImplBase.sendMessageBlock(WsRemoteEndpointImplBase.java:328)
	at org.apache.tomcat.websocket.WsRemoteEndpointImplBase.sendMessageBlock(WsRemoteEndpointImplBase.java:263)
	at org.apache.tomcat.websocket.WsSession.sendCloseMessage(WsSession.java:817)
	at org.apache.tomcat.websocket.WsSession.onClose(WsSession.java:740)
	at org.apache.tomcat.websocket.server.WsHttpUpgradeHandler.close(WsHttpUpgradeHandler.java:250)
	at org.apache.tomcat.websocket.server.WsHttpUpgradeHandler.upgradeDispatch(WsHttpUpgradeHandler.java:164)
	at org.apache.coyote.http11.upgrade.UpgradeProcessorInternal.dispatch(UpgradeProcessorInternal.java:60)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:59)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:872)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1705)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.writev0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:51)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:182)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:130)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:493)
	at org.apache.tomcat.util.net.NioChannel.write(NioChannel.java:129)
	at org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper$NioOperationState.run(NioEndpoint.java:1597)
	at org.apache.tomcat.util.net.SocketWrapperBase$OperationState.start(SocketWrapperBase.java:1031)
	at org.apache.tomcat.util.net.SocketWrapperBase.vectoredOperation(SocketWrapperBase.java:1420)
	at org.apache.tomcat.util.net.SocketWrapperBase.write(SocketWrapperBase.java:1346)
	at org.apache.tomcat.util.net.SocketWrapperBase.write(SocketWrapperBase.java:1317)
	at org.apache.tomcat.websocket.server.WsRemoteEndpointImplServer.doWrite(WsRemoteEndpointImplServer.java:93)
	at org.apache.tomcat.websocket.WsRemoteEndpointImplBase.writeMessagePart(WsRemoteEndpointImplBase.java:513)
	at org.apache.tomcat.websocket.WsRemoteEndpointImplBase.sendMessageBlock(WsRemoteEndpointImplBase.java:315)
	... 14 common frames omitted

**Comment by vinokurig** [comment]
Could you please share the entire file, so I can see the steps which caused the error?

**Comment by mukulk2020** [comment]
[logs.txt](https://github.com/eclipse-che/che-theia/files/9682023/logs.txt)

**Comment by mukulk2020** [comment]
@svor @vinokurig  Please help me on this .

**Comment by vinokurig** [comment]
@mukulk2020 I don't see any factory related operations in the log file. Did you create the factory before generating the log file? 
The error above is not related to the factory.

**Comment by mukulk2020** [comment]
@svor @vinokurig  No , I didn't create any factory manually . I just installed the dev spaces through operator hub and trying to create a workspace that points to a private Git Repo ..

**Comment by vinokurig** [comment]
Sorry for confusing with factories, but did you create the workspace by the GitHub Url before generating the log file?

**Comment by mukulk2020** [comment]
yes ..

**Comment by vinokurig** [comment]
Ok, could you please delete the created workspace and start a new one. Then share a new log file please.

**Comment by mukulk2020** [comment]
@vinokurig As per the documentation https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-github/ , what would be the Application name and homepage URL ? 
image

Application URL : Che .
Homepage URL : https://<che_fqdn>/

what is this <che_fqdn> ?

**Comment by vinokurig** [comment]
The Application name is `Che` and the home page Url is the Eclipse-Che endpoint.

**Comment by mukulk2020** [comment]
[devspaces-748c648f7f-mr78f-devspaces (1).log](https://github.com/eclipse-che/che-theia/files/9696708/devspaces-748c648f7f-mr78f-devspaces.1.log)

@vinokurig Added the latest logs ..

**Comment by mukulk2020** [comment]
[devspaces-748c648f7f-mr78f-devspaces (2).log](https://github.com/eclipse-che/che-theia/files/9696776/devspaces-748c648f7f-mr78f-devspaces.2.log)
@vinokurig Please check the latest logfile ...

**Comment by mukulk2020** [comment]
I tried this approach [https://www.youtube.com/watch?v=jpb4tY-gqYE] as well , but still it is asking for git username/password . @vinokurig I am stuck on this . please help

**Comment by vinokurig** [comment]
Investigation in progress :)

**Comment by vinokurig** [comment]
@mukulk2020 
Can you manually retrieve the devfile:
```
curl "https://raw.githubusercontent.com/xxxx/xxx-openshift-pet-clinic-app/HEAD/.devfile.yaml" -H "Authorization: token <github token>"
``` 
where `github token` is either a GitHub personal access token or the authentication token from the `
personal-access-token-...` secret of the user namespace

**Comment by mukulk2020** [comment]
Yes this is working fine .
image

**Comment by vinokurig** [comment]
@mukulk2020 
Did you use the token from the `personal-access-token-...` secret?

**Comment by mukulk2020** [comment]
Was used perrsonal-access-token for running above curl command

**Comment by vinokurig** [comment]
Ok, could you please try the token from the ` perrsonal-access-token-...` secret of the user namespace?

**Comment by mukulk2020** [comment]
I am using the same token in our secret "https://mukulk2020:ghp_bP4UAFWXXXXXXXXX@github.com":

image

**Comment by mukulk2020** [comment]
It is asking for credentials ..
image

**Comment by vinokurig** [comment]
I see that the `mount-path` is `/home/theia/.git-credentials`, can you share the content of the `devworkspace-gitconfig` configmap of the user namespace?

**Comment by vinokurig** [comment]
This is a wrong configmap, do you have `devworkspace-gitconfig` configmap?

**Comment by mukulk2020** [comment]
image
image

**Comment by vinokurig** [comment]
I mean in the workpsce pod's details like:
screenshot-console-openshift-console apps cluster-9vv8t 9vv8t sandbox1227 opentlc com-2022 10 04-13_47_30

**Comment by vinokurig** [comment]
And in the workspace terminal can you share the output of the `git config -l` command?

**Comment by mukulk2020** [comment]
image
image

**Comment by vinokurig** [comment]
@amisevsk Any ideas why the `.git-credentials` file is not mounted to the workspace pod?

**Comment by amisevsk** [comment]
Is there a DWO git credentials secret present in the namespace? Looking at the screenshot from https://github.com/eclipse-che/che-theia/pull/727#issuecomment-1266611444, it looks like it's trying to be but is missing the DWO annotations.

To see which secrets the DevWorkspace Operator can see in the namespace, use the command `oc get secrets -l "controller.devfile.io/watch-secret"`. There should be at least one secret in your namespace, e.g.
```bash
$ oc get secrets -l 'controller.devfile.io/watch-secret,controller.devfile.io/git-credential'
apiVersion: v1
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
items:
- apiVersion: v1
  kind: Secret
  metadata:
    annotations:
      che.eclipse.org/automount-workspace-secret: "true"
      che.eclipse.org/che-userid: <redacted>
      che.eclipse.org/git-credential: "true"
      che.eclipse.org/mount-as: file
      che.eclipse.org/mount-path: /.git-credentials
      che.eclipse.org/scm-url: https://github.com
      che.eclipse.org/scm-username: amisevsk
      controller.devfile.io/mount-path: /.git-credentials
    creationTimestamp: "2022-09-23T17:56:58Z"
    labels:
      app.kubernetes.io/component: workspace-secret
      app.kubernetes.io/part-of: che.eclipse.org
      controller.devfile.io/git-credential: "true"
      controller.devfile.io/watch-secret: "true"
    name: git-credentials-secret-dxuec
    namespace: amisevsk-dev
    resourceVersion: "1536799218"
    uid: 7c3fe7ef-d62a-43ad-aa5e-53d580c033af
  type: Opaque
  data:
    credentials: <redacted>
```
In addition, the `.data.credentials` field should decode to something like `https://oauth2:gho_<token>@github.com` (`base64 -d`)

Note that the labels/annotations
```
metadata:
  annotations:
    controller.devfile.io/mount-path: /.git-credentials
  labels:
    controller.devfile.io/git-credential: "true"
    controller.devfile.io/watch-secret: "true"
```
are required.

**Comment by vinokurig** [comment]
@mukulk2020 Can you deploy upstream Che and try the flow there?

**Comment by mukulk2020** [comment]
@vinokurig @amisevsk @svor Thanks for your inputs . Workspace is not asking of any username/password of my private github repo after adding below secret in my workspace .

kind: Secret
apiVersion: v1
metadata:
  name: git-credentials-secret
  namespace: che-kube-admin-devspaces-9q25gu
  labels:
    app.kubernetes.io/component: workspace-secret
    app.kubernetes.io/part-of: che.eclipse.org
    controller.devfile.io/git-credential: 'true'
    controller.devfile.io/watch-secret: 'true'
  annotations:
    che.eclipse.org/automount-workspace-secret: 'true'
    che.eclipse.org/mount-as: file
    controller.devfile.io/mount-path: /home/theia/.git-credentials
data:
  credentials: >-
    aHR0cHM6Ly9tdWt1bGsyMDIwOmdocF9iUDRXXXXXXXXXXXXXXXXXXXXXXl0aHViLmNvbQ==
type: Opaque

Above issue is resolved . But I am still facing below 2 issues 

1) Workspace is not picked up my devfile that exist in private github repo and during creation of workspace it is saying that "Devfile could not be found "

image

2) When i try to add some plugin from IDE (view-plugin) , it is showing be below error

image


@vinokurig @amisevsk @svor  Please suggest .

**Comment by mukulk2020** [comment]
@olexii4 @akurinnoy  Please support me on this and let me know in which release you are planning to fix this if this a issue from dashboard side.

**Comment by mukulk2020** [comment]
@olexii4 @akurinnoy @amisevsk Please suggest.

**Comment by amisevsk** [comment]
@mukulk2020 If you're still seeing this problem please open a new issue in https://github.com/eclipse/che -- I don't think anyone is following this discussion (it's a 2+ year old PR now :slightly_smiling_face:)
